### PR TITLE
Add simplification rules for the `CONCAT` function

### LINKED ISSUE #3683: Concatenate contiguous literal arguments of `CONCAT` during the expression simplification.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
For an expression `concat(col1, 'a', 'b', col2, 'd', 'f', ...)` we can simplify it to `concat(col1, 'ab', col2, 'df` ...)`

**Describe the solution you'd like**
1. Ignore all `null` literals
2. cast all literals to `UTF8`
3. concatenate contiguous literals.

**Describe alternatives you've considered**
We could not do this.

**Additional context**

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

# Which issue does this PR close?

Closes #3683 .

 # Rationale for this change

# What changes are included in this PR?

1. cast all scalar values to `utf8`
2. drop `null` values
3. concatenate contiguous scalar values

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 1]
```diff
@@ -849,12 +849,61 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
...
+                let mut contiguous_scalar = "".to_string();
+                for e in args {
+                    match e {
+                        // ignore `null` scalar and concatenate it with `contiguous scalar`.
+                        Expr::Literal(x) => {
+                            match x {
+                                // true --> '1', false --> '0'
+                                ScalarValue::Boolean(b) => {
+                                    contiguous_scalar += b
+                                        .map(|b| if b { "1" } else { "0" })
```
Comment (HaoYang670) [inline datafusion/optimizer/src/simplify_expressions.rs:867]: The behavior of `to_string` doesn't 100% match the `cast_to_utf8` kernel in `arrow-rs`. 
(BooleanType -> '1' / '0', and `null -> empty string)
Maybe we should reimplement it or add a `cast_to_utf8` function for `ScalarValue`.

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 2]
```diff
@@ -1263,6 +1312,37 @@ mod tests {
...
+    fn test_simplify_concat() {
+        fn build_concat_expr(args: &[Expr]) -> Expr {
+            Expr::ScalarFunction {
+                fun: BuiltinScalarFunction::Concat,
+                args: args.to_vec(),
+            }
+        }
+
+        let null = Expr::Literal(ScalarValue::Utf8(None));
+        // concat(true, c0, false, null, 'hello', c1, 12, 3.4) --> concat('1', c0, '0hello', c1, '123.4')
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:1325]: For example, in this case the true, false, etc will already be cast to utf8 by the time simplify sees this:

```
Plan("'datafusion.public.values' not found")
❯ create table foo as values (1);
0 rows in set. Query took 0.000 seconds.
❯ select * from foo
;
+---------+
| column1 |
+---------+
| 1       |
+---------+
1 row in set. Query took 0.001 seconds.
❯ explain select concat(true, false, 'bar', column1) from foo;
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                 |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: concat(Boolean(true), Boolean(false), Utf8("bar"), #foo.column1)                                                                                         |
|               |   TableScan: foo projection=[column1]                                                                                                                                |
| physical_plan | ProjectionExec: expr=[concat(CAST(true AS Utf8), CAST(false AS Utf8), bar, CAST(column1@0 AS Utf8)) as concat(Boolean(true),Boolean(false),Utf8("bar"),foo.column1)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                                                                      |
|               |                                                                                                                                                                      |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set. Query took 0.002 seconds.
```

File: datafusion/optimizer/src/simplify_expressions.rs (hunk 1 above)
Comment (alamb) [reply datafusion/optimizer/src/simplify_expressions.rs:867]: I agree finding some way to make the behavior the same is important if datafusion will rewrite the expression

I think in general, we should not try to reimplement the behavior in the simplify logic. By the time an expresison gets here, it should have been cast to `Utf8` anyways -- by the type coercion logic. 

I would recommend only simplifying calls where all the arguments are ScalarValue::Utf8

File: datafusion/optimizer/src/simplify_expressions.rs (hunk 2 above)
Comment (HaoYang670) [reply datafusion/optimizer/src/simplify_expressions.rs:1325]: Should we file another issue to add type-coercion rules for `concat` and `concat_ws`?

**Comment by HaoYang670** [comment]
Mark the PR as a draft because the type-coercion rule for `concat` hasn't been added yet.

**Comment by HaoYang670** [comment]
File #3720 to add type coercion rule.

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 3]
```diff
@@ -1352,6 +1396,30 @@ mod tests {
...
+            null.clone(),
+            col("c0"),
+            lit("hello "),
+            null.clone(),
+            lit("rust"),
+            col("c1"),
+            lit(""),
+            null,
+        ]);
+        let expected = build_concat_expr(&[col("c0"), lit("hello rust"), col("c1")]);
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:1419]: this is very clever

**Review by alamb — APPROVED** [review]
This is great -- thanks @HaoYang670

**Pull request merged**